src/System.Buffers.Experimental/System/Buffers/BufferExtensions.cs

```diff
@@ -122,103 +122,203 @@ internal static int IndexOfStraddling(this ReadOnlySpan<byte> first, IReadOnlyMe
         static readonly int s_longSize = Vector<ulong>.Count;
         static readonly int s_byteSize = Vector<byte>.Count;
 
-        public static int IndexOfVectorized(this Span<byte> buffer, byte value)
+        public unsafe static int IndexOfVectorized(this Span<byte> buffer, byte value)
         {
-            Debug.Assert(s_longSize == 4 || s_longSize == 2);
+            var index = -1;
+            var length = buffer.Length;
+            if (length == 0)
+            {
+                goto exit;
+            }
 
-            var byteSize = s_byteSize;
+            fixed (byte* pHaystack = &buffer.DangerousGetPinnableReference())
+            {
+                var haystack = pHaystack;
```

Comment:
Could you share code between Span and ReadOnlySpan implementations after getting the pointer?